### PR TITLE
ISPN-15186 SIFS may not shut down properly with compactor still running

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-15186
https://issues.redhat.com/browse/ISPN-15196
https://issues.redhat.com/browse/ISPN-15222

Also includes the issue where the compactor can attempt to compact an entry that does not have an index yet due to timing.

### CODE REVIEW

**Comment by wburns** [comment]
Also to note the StressTest has passed multiple times for me now without any issues. Note that tracing on or off can mess with timeout.

**Review by jabolina — APPROVED** [review]
LGTM. Pretty nice.

**Comment by wburns** [comment]
Yes, good catch. Will update on Monday.

File: core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java [hunk 1]
```diff
@@ -49,21 +73,135 @@ protected Configuration buildConfig(ConfigurationBuilder configurationBuilder) {
...
+      processors[0] = original;
+      queue.forEach(original::onNext);
+
+      // Now expiration compaction should finish as indexing will occur
+      assertTrue(latch.await(10, TimeUnit.SECONDS));
+
+      // The stop method blocks for NBSIS as we are faking the blocking manager
+      Future<CompletionStage<Void>> stage = fork(() -> store.stop());
+
+      eventually(stage::isDone);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java]: Updated.

**Comment by jabolina** [comment]
Seems CI crashed:

```text
14:27:03  13:27:03.146 [ERROR] Errors: 
14:27:03  13:27:03.146 [ERROR]   PessimisticTxPartitionAndMergeDuringPrepareTest.testDegradedPartition:29->BasePessimisticTxPartitionAndMergeTest.doTest:49->BaseTxPartitionAndMergeTest.createKeys:131 » IllegalState Could not find any segment owned by Cache 'pes-cache'@PessimisticTxPartitionAndMergeDuringPrepareTest-NodeC, [Cache 'pes-cache'@PessimisticTxPartitionAndMergeDuringPrepareTest-NodeB], primary segments: {33-44 48-59 153-174 185-194 218 234-235 242-244}, backup segments: {Cache 'pes-cache'@PessimisticTxPartitionAndMergeDuringPrepareTest-NodeB={0-11 22-23 74-78 95-105 112-113 127-130 136-146 206-217 249-250 253-255}}
14:27:03  13:27:03.146 [INFO] 
...
14:27:03  13:27:03.157 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project infinispan-core: 
14:27:03  13:27:03.157 [ERROR] 
14:27:03  13:27:03.157 [ERROR] Please refer to /home/infinispan/workspace/Infinispan_PR-11338/core/target/surefire-reports for the individual test results.
14:27:03  13:27:03.157 [ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
14:27:03  13:27:03.157 [ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
```

Going to replay.

**Comment by jabolina** [comment]
@wburns The SFSToSIFSMigratorTest are related?

**Comment by tristantarrant** [comment]
@wburns did you have a look at the failures ?

**Comment by wburns** [comment]
It is related, underlying cause is:
```
09:38:36,958 ERROR (blocking-thread-SFSToSIFSMigratorTest-NodeA-p13-t2:[]) [o.i.p.s.Index] Error encountered with index, SIFS may not operate properly.
java.lang.AssertionError: null
	at org.infinispan.persistence.sifs.IndexNode.copyWith(IndexNode.java:780) ~[classes/:?]
	at org.infinispan.persistence.sifs.IndexNode.setPosition(IndexNode.java:418) ~[classes/:?]
	at org.infinispan.persistence.sifs.Index$Segment.accept(Index.java:563) ~[classes/:?]
	at org.infinispan.persistence.sifs.Index$Segment.accept(Index.java:456) ~[classes/:?]
	at io.reactivex.rxjava3.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:65) ~[rxjava-3.1.4.jar:?]
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableObserveOn$ObserveOnSubscriber.runAsync(FlowableObserveOn.java:404) ~[rxjava-3.1.4.jar:?]
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableObserveOn$BaseObserveOnSubscriber.run(FlowableObserveOn.java:178) ~[rxjava-3.1.4.jar:?]
	at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:324) ~[rxjava-3.1.4.jar:?]
	at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker.runEager(ExecutorScheduler.java:289) ~[rxjava-3.1.4.jar:?]
	at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:250) ~[rxjava-3.1.4.jar:?]
	at org.infinispan.util.concurrent.BlockingManagerImpl.lambda$asExecutor$9(BlockingManagerImpl.java:336) ~[classes/:?]
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18) [jboss-threads-3.5.0.Final.jar:3.5.0.Final]
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513) [jboss-threads-3.5.0.Final.jar:3.5.0.Final]
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538) [jboss-threads-3.5.0.Final.jar:3.5.0.Final]
	at java.lang.Thread.run(Thread.java:833) [?:?]
```

**Comment by wburns** [comment]
So this is actually a bug in the SFSToSIFSStore migration class that is exposed in this PR. 

https://github.com/infinispan/infinispan/blob/b28d3509a72f5bf62229034927e31b9bbc300c55/core/src/main/java/org/infinispan/configuration/parsing/SFSToSIFSStore.java#L67

This line is accidentally taking the segment from the MarshallableEntry instead of the key. The odd thing is I am not sure why the test passed before, it must have ignored the segment when doing compaction... I will look a bit closer.

```
09:40:54,686 TRACE (blocking-thread-SFSToSIFSMigratorTest-NodeA-p8-t4:[]) [o.i.p.s.Index] Indexing IndexRequest{key=key, serializedKey=ByteBufferImpl{length=13, offset=0, bytes=9801028A01070A05\J03\k\e\y (13 bytes)}, cacheSegment=213, file=0, offset=0, prevFile=-1, prevOffset=-1, size=55, type=UPDATE}
09:40:54,696 TRACE (blocking-thread-SFSToSIFSMigratorTest-NodeA-p8-t4:[]) [o.i.p.s.Index] Indexing IndexRequest{key=key, serializedKey=ByteBufferImpl{length=13, offset=0, bytes=9801028A01070A05\J03\k\e\y (13 bytes)}, cacheSegment=213, file=0, offset=55, prevFile=-1, prevOffset=-1, size=55, type=UPDATE}
09:40:58,136 TRACE (blocking-thread-SFSToSIFSMigratorTest-NodeA-p13-t2:[]) [o.i.p.s.Index] Indexing IndexRequest{key=key, serializedKey=ByteBufferImpl{length=13, offset=0, bytes=9801028A01070A05\J03\k\e\y (13 bytes)}, cacheSegment=99, file=0, offset=55, prevFile=0, prevOffset=0, size=-1, type=DROPPED}
```
As you can see from the TRACE the cacheSegment doesn't match, it is supposed to be 99

**Comment by wburns** [comment]
Test should be fixed with inclusion of fix for [ISPN-15222](https://issues.redhat.com/browse/ISPN-15222)

File: core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java [hunk 2]
```diff
@@ -41,6 +64,7 @@ protected Configuration buildConfig(ConfigurationBuilder configurationBuilder) {
             .addSoftIndexFileStore()
             .dataLocation(Paths.get(tmpDirectory, "data").toString())
             .indexLocation(Paths.get(tmpDirectory, "index").toString())
+            .indexSegments(1)
```
Comment (jabolina) [inline core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java:67]: This one needed to make the index delay?

**Review by jabolina — APPROVED** [review]
Seems the tests are passing

File: core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java (hunk 2 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java:67]: No, it isn't required, just makes the test easier to diagnose and trace

**Comment by jabolina** [comment]
Thanks, @wburns!

**Pull request merged**